services/flake_detection.py

```diff
@@ -1,67 +1,170 @@
 from collections import defaultdict
-from dataclasses import dataclass, field
+from dataclasses import dataclass
 from datetime import datetime, timedelta
-from enum import Enum
+from functools import cached_property
 from logging import getLogger
 from sys import getsizeof
-from typing import List, Set, Tuple
 
 from sentry_sdk import metrics, trace
 from test_results_parser import Outcome
 
+from database.enums import FlakeSymptomType
 from database.models.core import Commit, Repository
 from database.models.reports import CommitReport, Test, TestInstance, Upload
 
 log = getLogger(__name__)
 
+FlakeDetectionResult = dict[Test, set[FlakeSymptomType]]
 
-@dataclass
-class FlakeDetectionObject:
-    counter: int = 0
-    branch: Set[str] = field(default_factory=set)
-    time: List[datetime] = field(default_factory=list)
 
+class BaseSymptomDetector:
+    """
+    blueprint class for a single symptom of flakiness detector
 
-class FlakeType(Enum):
-    FAILED_IN_DEFAULT_BRANCH = "failed_in_default_branch"
-    CONSECUTIVE_DIFF_OUTCOMES = "consecutive_diff_outcomes"
-    UNRELATED_MATCHING_FAILURES = "unrelated_matching_failures"
+    classes that implement this interface are expected to:
+    - maintain their own state necessary to detect a symptom of flakiness
+    - implement the ingest method that:
+        - receives a TestInstanceInfo object
+        - adds the necessary information to its state to be able to
+          detect the symptom of flakiness it is responsible for
+    - implement the detect method that:
+        - Runs the detection based on all the instances that it has
+          ingested
+        - returns an InstanceMapping
+    """
 
+    def ingest(self, instance):
+        raise NotImplementedError()
 
-@dataclass
-class FlakeDetectionContext:
-    is_curr_flake = False
-    curr_test_id = None
-    flake_dict = defaultdict(FlakeDetectionObject)
-    commit_to_outcome = dict()
+    def detect(self) -> dict[str, TestInstance]:
+        raise NotImplementedError()
 
-    def reset(self, test_id):
-        self.curr_test_id = test_id
-        self.is_curr_flake = False
-        self.flake_dict.clear()
-        self.commit_to_outcome.clear()
+    @cached_property
```

Comment:
Let's remember that having a class-wide cached property is not a good idea cause the cache is shared among all instances of the class... I think that was the gotcha.

I do think that each of these classes should return a single `FlakeSymtomType` so this seems fine. Probably it's what you intended, but just stating it again

Reply:
That being said, I don't think we gain anything by having this cached.... in all your implementations it's just a constant being returned. Maybe I'd remove this cached_property, but nit in any case.